Hazel/Udp/UdpConnection.Fragmented.cs

```diff
@@ -0,0 +1,241 @@
+
+                WriteBytesToConnection(buffer, buffer.Length);
+            }
+        }
+
+        protected void FragmentMessageReceive(MessageReader messageReader)
+        {
+            if (ProcessReliableReceive(messageReader.Buffer, 1, out var id))
+            {
+                messageReader.Position += 3;
+
+                var fragmentsCount = messageReader.ReadUInt16();
+                var fragmentedMessageId = messageReader.ReadUInt16();
+
+                lock (_fragmentedMessagesReceived)
+                {
+                    if (!_fragmentedMessagesReceived.TryGetValue(fragmentedMessageId, out var fragmentedMessage))
+                    {
+                        _fragmentedMessagesReceived.Add(fragmentedMessageId, fragmentedMessage = new FragmentedMessage(fragmentsCount));
+                    }
+
+                    var buffer = new byte[messageReader.Length - messageReader.Position];
+                    Buffer.BlockCopy(messageReader.Buffer, messageReader.Position, buffer, 0, messageReader.Length - messageReader.Position);
+
+                    fragmentedMessage.Fragments.Add(new FragmentedMessage.Fragment(id, buffer));
+
+                    if (fragmentedMessage.Fragments.Count == fragmentsCount)
+                    {
+                        var reconstructed = fragmentedMessage.Reconstruct();
+                        InvokeDataReceived(SendOption.Reliable, MessageReader.Get(reconstructed), 1, reconstructed.Length);
+
+                        _fragmentedMessagesReceived.Remove(id);
+                    }
+                }
+            }
+        }
+
+        protected class FragmentedMessage
+        {
+            /// <summary>
+            ///     The total number of fragments expected.
+            /// </summary>
+            public int FragmentsCount { get; }
+
+            /// <summary>
+            ///     The fragments received so far.
+            /// </summary>
+            public HashSet<Fragment> Fragments { get; } = new HashSet<Fragment>();
+
+            public byte[] Reconstruct()
+            {
+                if (Fragments.Count != FragmentsCount)
+                {
+                    throw new HazelException("Can't reconstruct a FragmentedMessage until all fragments are received");
+                }
+
+                var buffer = new byte[Fragments.Sum(x => x.Data.Length)];
+
+                var offset = 0;
+                foreach (var fragment in Fragments.OrderBy(fragment => fragment.Id))
+                {
+                    var data = fragment.Data;
+                    Buffer.BlockCopy(data, 0, buffer, offset, data.Length);
+                    offset += data.Length;
+                }
+
+                return buffer;
+            }
+
+            public FragmentedMessage(int fragmentsCount)
+            {
+                FragmentsCount = fragmentsCount;
+            }
+
+            public readonly struct Fragment
```

Comment:
Fun fact although it doesn't matter if you use an array up there, but: Because structs are byval equality and you're putting a new byte[] in each one, no two will ever be equal (even if both byte[] contains the same bytes) so you weren't really getting anything out of the HashSet and a List woulda been faster.